File: kickstart2020/A/Allocation/source.cpp
#include<iostream>
using namespace std;

typedef long long int ll;
typedef pair<ll,ll> pii;
#define REP(i,n) for(ll i=0;i<n;i++)

#ifdef DEBUG
#define dbg(x) x
#define dbgp(x) cerr << x << endl;
#else
#define dbg(x) //x
#define dbgp(x) //cerr << x << endl;
#endif

int solve() {
    int n, b;
    cin >> n >> b;

    int xs[n];
    REP(i, n) {
        cin >> xs[i];
    }

    sort(xs, xs+n);

    int count = 0;
    int ba = 0;
    REP(i, n) {
        if (ba + xs[i] > b) {
            break;
        }
        count++;
        ba += xs[i];
    }
    return count;
}

int main(){
    int t;
    cin >> t;
    REP(i, t) {
        cout << "Case #" << (i+1) << ": ";
        cout << solve() << endl;
    }
}

/**
        (answer, _) = foldl' folder (0,0) sorted 
        sorted = sort xs
        folder :: (IInt, IInt) -> IInt -> (IInt, IInt)
        folder (count, total) x
            | total + x > b = (count, total+x) 
            | otherwise = (count+1, total+x)
            */

File: kickstart2020/A/Allocation/source.hs
import Data.List
import Data.Int
import qualified Data.Array.Unboxed as A
import Data.Array((!))
import Numeric

type IInt = Int64

fastRead :: String -> IInt
fastRead' ('-':s) = case readDec s of [(n, "")] -> -n
fastRead' s = case readDec s of [(n, "")] -> n

fastRead str = answer `seq` answer
    where
        answer = fastRead' str

main = interact solve

type Input = (IInt,[IInt])  -- B,Ai
type Output = IInt

solve :: String -> String
solve input = intercalate "\n" formattedOutputs
    where
        _:xs = words input
        inputs = inputParser xs
        outputs = map solver inputs
        formattedOutputs = map outputFormatter (zip [1..] outputs)

inputParser :: [String] -> [Input]
inputParser = inputParser' . map fastRead

inputParser' :: [IInt] -> [Input]
inputParser' [] = []
inputParser' (n:b:xs) = xs `seq` (b, as):(inputParser' rest)
    where
        (as,rest) = splitAt (fromIntegral n) xs

outputFormatter :: (Int,Output) -> String
outputFormatter (idx,x) = x `seq` "Case #" ++ (show idx) ++ ": " ++ (show x)

solver :: Input -> Output
solver (b,xs) = answer `seq` answer
    where
        (answer, _) = foldl' folder (0,0) sorted 
        sorted = sort xs
        folder :: (IInt, IInt) -> IInt -> (IInt, IInt)
        folder (count, total) x
            | total + x > b = (count, total+x) 
            | otherwise = (count+1, total+x)
